caffe2/serialize/intermediate_model.h

```diff
@@ -0,0 +1,395 @@
+    return dataType_;
+  }
+
+  const std::vector<int64_t>& dims() const {
+    return dims_;
+  }
+
+  std::shared_ptr<SharedData> data() const {
+    return data_;
+  }
+
+  int64_t offset() const {
+    return offset_;
+  }
+
+  const IntermediateDeviceOption& deviceDetail() const {
+    return deviceDetail_;
+  }
+
+  const std::vector<int64_t>& strides() const {
+    return strides_;
+  }
+
+  bool noContent() const {
+    return noContent_;
+  }
+
+
+  void setData(std::shared_ptr<SharedData> data) {
+    AT_ASSERTM(!noContent_, "noContent_ is true, but set content!");
+    data_ = data;
+  }
+
+  void setDataType(int64_t data_type) {
+    dataType_ = data_type;
+  }
+
+  void setDims(const std::vector<int64_t>& dims) {
+    dims_ = dims;
+  }
+
+  void setOffset(int64_t offset) {
+    offset_ = offset;
+  }
+
+  void setStrides(const std::vector<int64_t>& strides) {
+    strides_ = strides;
+  }
+
+  void setDeviceDetail(const IntermediateDeviceOption& device_detail) {
+    deviceDetail_ = device_detail;
+  }
+
+ private:
+  std::string name_;
+  int64_t dataType_;
+  std::vector<int64_t> dims_;
+  int64_t offset_ = 0;
+  std::vector<int64_t> strides_;
+  // TODO: since we still have 2 different Tensor classes in Caffe2 and PyTorch
+  // right now, let's just store the data pointer, and create Tensors
+  // while converting IntermediateModel to the JitScriptModule/Predictor/etc.
+  std::shared_ptr<SharedData> data_;
+  IntermediateDeviceOption deviceDetail_;
+  bool noContent_ = false;
+};
+
+class CAFFE2_API IntermediateParameter final {
+ public:
+  // constructors
+  IntermediateParameter() = default;
+
+  explicit IntermediateParameter(const std::string& name, bool is_buffer, bool require_gradient) :
+    name_(name), isBuffer_(is_buffer), requireGradient_(require_gradient) {}
+
+  explicit IntermediateParameter(torch::ParameterDef* param_def,
+      std::unordered_map<uint64_t, std::shared_ptr<SharedData>>* id_data,
+      DeserializeMode mode);
+
+  // dump data to ParameterDef, invoked in serialize
+  void dump(torch::ParameterDef* param_def);
+
+  // getters/setters
+  const std::string& name() const {
+    return name_;
+  }
+
+  bool isBuffer() const {
+    return isBuffer_;
+  }
+
+  bool requireGradient() const {
+    return requireGradient_;
+  }
+
+  const IntermediateTensor& tensor() const {
+    return tensor_;
+  }
+
+  IntermediateTensor* mutableTensor() {
+    return &tensor_;
+  }
+
+  void setName(const std::string& name) {
+    name_ = name;
+  }
+
+  void setIsBuffer(bool is_buffer) {
+    isBuffer_ = is_buffer;
+  }
+
+  void setRequireGradient(bool require_gradient) {
+    requireGradient_ = require_gradient;
+  }
+
+ private:
+  std::string name_;
+  bool isBuffer_ = false;
+  bool requireGradient_ = false;
+  IntermediateTensor tensor_;
+};
+
+class CAFFE2_API IntermediateMethod final {
+ public:
+  // constructors
+  IntermediateMethod() = default;
+
+  explicit IntermediateMethod(torch::MethodDef* method_def);
```

Comment:
so, explicit copy/move constructors seems to work - https://godbolt.org/z/a11-CB

but, actually  even if you  don't specify them, compiler generates the write wrappers for you https://godbolt.org/z/Ej6QGg

Reply:
If I delete assignment operator, it doesn't work any more: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/28083/console

Reply:
That's because you did this:

```
  methods->resize(1);
```

Don't do that. Resize requires copy constructor/assignment and you can't implement this correctly on a class with a `unique_ptr`. Do a `reserve` instead and emplace it later.